gateways article part 2 (#435)
ipfs/ipfs-blog: src/_blog/practical-explainer-ipfs-gateways-part-2.md

---
title: A Practical Explainer for IPFS Gateways - Part 2
description: 'Learn tips and tricks for using IPFS gateways in real-world applications - debugging, performance and more'
author: Daniel Norman
date: 2022-06-09
permalink: '/2022-06-09-practical-explainer-ipfs-gateways-2/'
translationKey: ''
header_image: '/2022-ipfs-gateways-2.png'
tags:
  - gateways
  - debugging
  - DHT
  - content routing
---

The Interplanetary File System (IPFS) is a peer-to-peer protocol for storing and accessing **files and websites**. As a distributed **peer-to-peer** protocol, it's fundamentally different from the HTTP protocol. With the help of IPFS Gateways, it's possible to reap the benefits of IPFS using the HTTP protocol.

This blog post is the second of a two-part series about IPFS gateways:

> [The first part](https://blog.ipfs.io/2022-06-09-practical-explainer-ipfs-gateways-1/) was mostly theoretical, covering the principles behind IPFS, the challenges with the client-server model, how IPFS approaches these challenges with peer-to-peer networking and content addressing, the relationship between IPFS and HTTP(S), and finally an introduction to IPFS HTTP gateways and resolution styles.

Author: Changing to a bullet list

From a high level, when faced with challenges (either slowness or timeouts) fetching a CID from an IPFS gateway, it's typically related to one of the following:

- The IPFS gateway
- The provider of the CID, i.e. the IPFS node pinning the CID might be unreachable or down.
- You (or the pinning service) are not providing your CIDs to the IPFS network. Providing is the process by which providers of a given CID advertise it to the distributed hash table (DHT) to make it discoverable.
- Network latency between the client and the IPFS gateway or the gateway and the provider.

Given all of these factors, it's difficult giving blanket advice. This is where understanding the lifecycle of a CID request to an IPFS gateway is useful as it will allow debugging problems quickly.

## IPFS gateway request lifecycle

When a request for a CID reaches an IPFS gateway, the gateway checks whether the CID is cached locally before attempting to retrieve it from the network.

If the CID is in the gateway's cache, the gateway will respond to the HTTP request with the CID's content.

> > **Note:** Cache here can be either an HTTP cache or the local caching by the IPFS node.

Author: I like "local data store". It makes the distinction clear! Thanks

![How nftstorage.link works](../assets/nftstoragelink-gateway.png)

[NFT.Storage Gateway SuperHot perma-cache](https://nft.storage/blog/post/2022-05-24-superhot-gateway-announcement/) is a paid feature recently launched by the NFT.Storage team, which is similar to dedicated gateways. It gives you the ability to preload your CIDs in all of Cloudflare's 270 points of presence, giving a lightning-fast read experience via the closest CDN location to your users

Finally, a **self-hosted gateway** refers to an IPFS node(s) configured as a gateway that is hosted by you, either on your local machine or in the cloud.

Choosing from the three approaches depends on your requirements, if performance is critical, self-hosting an IPFS node and gateway or using a dedicated gateway is a reasonable choice that can complement the usage of public gateways.

## Best practices for self-hosting an IPFS node/gateway

If you are running an IPFS node that is also configured as an IPFS gateway, there are steps you can take to improve the discovery and retrievability of your CIDs.

- Set up [peering](https://docs.ipfs.io/how-to/peering-with-content-providers/) with the pinning services that pin your CIDs.

> - Ensure that you are correctly returning HTTP cache headers to the client if the IPFS gateway node is behind a reverse-proxy

Author: For sure!
I still need to learn more about how `X-Ipfs-Roots` can help with smarter caching.

> - Ensure that you are correctly returning HTTP cache headers to the client if the IPFS gateway node is behind a reverse-proxy

Reviewer: It was introduced in https://github.com/ipfs/go-ipfs/pull/8720 and is available in go-ipfs 0.13 already.
FWIW I included it in HTTP Gateway specs PR (https://github.com/ipfs/specs/pull/283) – [check markdown preview](https://github.com/ipfs/specs/blob/feat/gateway-specs/http-gateways/PATH_GATEWAY.md#x-ipfs-roots-response-header) (lmk of that PR if this description could be improved) :pray:

## Tip: Use a custom domain that you control as your IPFS gateway

Imagine the following scenario: you deploy your web app to IPFS which contains media with absolute URLs to a public gateway experiencing an outage. For example, your web app displays the image with an absolute path to the CID: https://bafybeibml5uieyxa5tufngvg7fgwbkwvlsuntwbxgtskoqynbt7wlchmfm.ipfs.infura-ipfs.io.

You may be able to reach your media using a different gateway, but since the web app's content is immutable, the links to the IPFS gateway which is down will not load.

For these reasons, it's sensible to use a domain within your control to route HTTP traffic to a gateway. This approach potentially gives you the flexibility to implement additional performance optimizations.

Practically speaking, this can be implemented using several approaches depending on your willingness to run infrastructure:

- Point a domain you control, e.g., `*.ipfs.yourdomain.io` point to a reverse proxy like nginx which will proxy requests to a public gateway, allowing you to switch public gateways if there's downtime.

> - Use [Cloudflare workers](https://workers.cloudflare.com/) to implement a lightweight proxy to IPFS gateways.

Author: Indeed. Might be worth mentioning https://www.fastly.com/products/edge-compute even though I haven't used it yet.

```
ipfs id -f '<addrs>' [PEER_ID]
```

The result of this command looks as follows:

```
/ip4/145.40.90.155/tcp/4001/p2p/12D3KooWSH5uLrYe7XSFpmnQj1NCsoiGeKSRCV7T5xijpX2Po2aT
/ip4/145.40.90.155/tcp/4002/ws/p2p/12D3KooWSH5uLrYe7XSFpmnQj1NCsoiGeKSRCV7T5xijpX2Po2aT
/ip6/2604:1380:45e1:2700::d/tcp/4001/p2p/12D3KooWSH5uLrYe7XSFpmnQj1NCsoiGeKSRCV7T5xijpX2Po2aT
/ip6/2604:1380:45e1:2700::d/tcp/4002/ws/p2p/12D3KooWSH5uLrYe7XSFpmnQj1NCsoiGeKSRCV7T5xijpX2Po2aT
```

To check if the CID is retrievable from the returned addresses, copy one of the public addresses and the CID into [IPFS Check](https://check.ipfs.network/):

![IPFS Check demo](../assets/ipfs-check-demo.gif)

IPFS Check will test whether the peer/node is dialable and whether the CID is retrievable from it.

### Debugging with pl-diagnose

[pl-diagnose](https://pl-diagnose.on.fleek.co/#/diagnose/access-content) is an alternative web-based tool that can perform a lot of the steps above directly from the browser.

You case use it to look up providers for a CID, checking if a [multiaddr](https://multiformats.io/multiaddr/) is reachable by other peers, and whether a node is serving the CID.

## Content publishing lifecycle

> Now that you're familiar with content retrieval, we'll take a look at the other end of retrieval, namely **content publishing** – the process by which content becomes discoverable in the IPFS network by advertising provider records.

Reviewer 2: ```suggestion
Now that you're familiar with content retrieval, we'll take a look at the other end of retrieval, namely **content publishing**. Content publishing is how your content becomes discoverable to peers in the IPFS network.
```
simplify sentences by breaking into two, use active voice

Author: Reads much better with shorter sentences and active voice

> **Note:** The term **publishing** is a bit misleading, because it refers to the publishing of provider records to the DHT; not the actual content. For this reason, you might see the term **advertising** also used.

> This is a continuous process that starts when you first add content to an IPFS node and repeats every 12 hours (by default) as long as the node is running.

Reviewer 2: I think this can be moved into line 150, after the list (or removed altogether as there's some overlap with #3).

> When you add a file to an IPFS node using the `ipfs add` command, the process can be broken down from a high level into the following steps:

Reviewer 2: ```suggestion
When you add a file to an IPFS node using the `ipfs add` command, here's how it becomes published and discoverable across the network:
```

1. The file is chunked into blocks and a [Merkle DAG](https://docs.ipfs.io/concepts/merkle-dag/) is constructed. You get back the root CID of the DAG.
2. The blocks of the file are made available over [Bitswap](https://docs.ipfs.io/concepts/bitswap/) so any peers can request.
3. Mappings of CIDs to network addresses (including CIDs of the block and the root CID) are advertised to the DHT. These **provider records** have an expiry time of 24 hours (accounting for provider churn) and need to be **reprovided** by the node every 12 hours (accounting for peer churn).

## Debugging content publishing

IPFS network measurements conducted by the [ProbeLab](https://blog.ipfs.io/2022-06-15-probelab/), show that [_content publishing is a bottleneck_](https://youtu.be/75ewjnT6B9Y?t=115) in IPFS. While there are efforts explored in that talk to improve this, it's useful to understand how to troubleshoot problems related to content publishing.

Generally speaking, as you add more files to your IPFS node, the longer reprovide runs take.

> 

Reviewer 2: ```suggestion
```
group these 2 sentences into a single paragraph since they're so closely related?

Author: Makes sense

**When a reprovide run takes longer than 24 hours (the expiry time for provider records), your CIDs will become undiscoverable**.

To find out how long a reprovide run takes, run the following command:

```sh
ipfs stats provide
```

The result should look like this:

```
TotalProvides:          7k (7,401)
AvgProvideDuration:     271.271ms
LastReprovideDuration:  13m16.104781s
LastReprovideBatchSize: 1k (1,858)
```

> If you notice that the `LastReprovideDuration` value is reaching close to 24 hours, you should consider one of the following options as a resolution:

Reviewer 2: ```suggestion
If you notice that the `LastReprovideDuration` value is reaching close to 24 hours, you should consider one of the following options:
```
